code/obj/item/gun/gun_parent.dm

```diff
@@ -457,3 +509,69 @@ var/list/forensic_IDs = new/list() //Global list of all guns, based on bioholder
 /obj/item/gun/proc/set_current_projectile(datum/projectile/newProj)
 	src.current_projectile = newProj
 	SEND_SIGNAL(src, COMSIG_GUN_PROJECTILE_CHANGED, newProj)
+
+/obj/item/gun/proc/do_camera_recoil(mob/user, turf/start, turf/target, POX, POY)
+	// calculate the mob's position relative to the target location
+	// this is backwards so that the output angle is the angle we knock the camera back
+	var/x_diff = (start.x - target.x) * world.icon_size - POX
+	var/y_diff = (start.y - target.y) * world.icon_size - POY
+
+	var/dir = arctan(x_diff, y_diff)
+	var/total_strength = src.recoil_strength
+	if (recoil_stacking_enabled)
+		total_strength += clamp(round(recoil_stacks) - recoil_stacking_safe_stacks,0,recoil_stacking_max_stacks) * recoil_stacking_amount
+	var/variance = clamp(total_strength * camera_recoil_sway_multiplier, camera_recoil_sway_min, camera_recoil_sway_max)
+
+	recoil_camera(user, dir, total_strength * camera_recoil_multiplier, variance)
+
+
+/obj/item/gun/proc/do_icon_recoil()
+	if (!icon_recoil_enabled)
+		return
+	while(src.recoil > 0)
+		var/timediff = TIME - recoil_last_shot
+		if (timediff >= recoil_reset)
+			recoil *= recoil_reset_mult
+			recoil -= recoil_reset_add // small linear part to aid with low values
+			recoil = clamp(recoil,0, recoil_max)
+			recoil_stacks = clamp(round(recoil_stacks) - 0.25, 0, recoil_stacking_max_stacks)
+
+		var/matrix/M = src.transform
+		var/jitter = round(recoil/10)
+		var/jittervalue = rand(-jitter, jitter)
+		if (src.recoil < current_anim_recoil || timediff > 0.2 DECI SECONDS) // stop the gun jerking up after you stop shooting
+			jittervalue = 0
+		var/target_recoil = round((recoil/recoil_max)*icon_recoil_cap) + jittervalue
+
+		var/recoil_diff = (current_anim_recoil - target_recoil)
+		current_anim_recoil = target_recoil
+		animate(src, transform = matrix(M, recoil_diff, MATRIX_ROTATE | MATRIX_MODIFY), 0.1)
+		sleep(0.1 SECONDS)
+	recoil_stacks = 0
+
+/obj/item/gun/proc/handle_recoil(mob/user, turf/start, turf/target, POX, POY, first_shot = TRUE)
+	if (!recoil_enabled)
+		return
+	var/start_recoil = FALSE
+	if (recoil == 0)
+		start_recoil = TRUE // if recoil is 0, make sure do_recoil starts
+
+	// Add recoil
+	var/stacked_recoil = 0
+	if (recoil_stacking_enabled)
+		recoil_stacks += 1
+		stacked_recoil = clamp(round(recoil_stacks) - recoil_stacking_safe_stacks,0,recoil_stacking_max_stacks) * recoil_stacking_amount
+
+	recoil += (recoil_strength + stacked_recoil)
+	recoil = clamp(recoil, 0, recoil_max)
+	recoil_last_shot = TIME
+	if (camera_recoil_enabled)
+		do_camera_recoil(user, start,target,POX,POY)
+
+	if (first_shot && src.current_projectile.shot_number > 1 && src.current_projectile.shot_delay > 0)
+		for (var/i=1 to src.current_projectile.shot_number-1)
+			SPAWN(i*src.current_projectile.shot_delay)
+				handle_recoil(user,start,target,POX,POY, FALSE)
```

Comment:
this feels very cursed

Reply:
it is. however projectile code (which handles burst) doesn't have a reference back to owner

Reply:
i could probably rework that (but i do have the big firemode seperation at about 75~80% completion, and this would definitely get in the way)